File: Adapter/cpp-source/TextStorage.cpp
//
// Created by Pavel Akhtyamov on 26/03/2018.
//

#include <fstream>
#include <sstream>
#include "TextStorage.h"


void TextStorage::Read(const std::string &filename) {
  std::ifstream stream(filename);

  std::string buffer;
  while (std::getline(stream, buffer)) {
    pairs_.push_back(ParseTokens(buffer));
  }
  stream.close();
}


void TextStorage::Write(const std::string &filename) {
  std::ofstream stream(filename);

  for (auto &pair : pairs_) {
    stream << pair.key << " " << pair.value << "\n";
  }
  stream.close();
}


KeyValuePair TextStorage::ParseTokens(const std::string &input_string) {
  std::stringstream tokenizer;
  tokenizer << input_string;

  std::string key;
  std::string value;

  tokenizer >> key >> value;
  return { key, value };
}


std::vector<KeyValuePair> TextStorage::GetPairs() const {
  return pairs_;
}

File: Adapter/python-source/TextStorage.py
import os
from collections import namedtuple

KeyValuePair = namedtuple('KeyValuePair', ['key', 'value'])


class TextStorage:
    def __init__(self):
        self.pairs: list[KeyValuePair] = []

    def read(self, filename):
        with open(filename) as fp:
            for line in fp:
                self.pairs.append(self.parse_tokens(line))

    def write(self, filename):
        with open(filename, 'w') as fp:
            fp.writelines(f'{pair.key} {pair.value}{os.linesep}' for pair in self.pairs)

    def parse_tokens(self, string: str) -> KeyValuePair:
        key, value = string.strip().split()

        return KeyValuePair(
            key=key,
            value=value
        )
